plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -25,22 +52,95 @@ namespace block_timing_util {
              fc::microseconds(cpu_effort_us * production_round_index);
    }
 
-   inline fc::time_point calculate_block_deadline(uint32_t cpu_effort_us, pending_block_mode mode, chain::block_timestamp_type block_time) {
-      const auto hard_deadline =
-                  block_time.to_time_point() - fc::microseconds(chain::config::block_interval_us - cpu_effort_us);
-      if (mode == pending_block_mode::producing) {
-         auto estimated_deadline = production_round_block_start_time(cpu_effort_us, block_time) + fc::microseconds(cpu_effort_us);
-         auto now                = fc::time_point::now();
-         if (estimated_deadline > now) {
-            return estimated_deadline;
-         } else {
-            // This could only happen when the producer stop producing and then comes back alive in the middle of its own
-            // production round. In this case, we just use the hard deadline.
-            return std::min(hard_deadline, now + fc::microseconds(cpu_effort_us));
+   inline fc::time_point calculate_producing_block_deadline(uint32_t cpu_effort_us, chain::block_timestamp_type block_time) {
+      auto estimated_deadline = production_round_block_start_time(cpu_effort_us, block_time) + fc::microseconds(cpu_effort_us);
+      auto now                = fc::time_point::now();
+      if (estimated_deadline > now) {
+         return estimated_deadline;
+      } else {
+         // This could only happen when the producer stop producing and then comes back alive in the middle of its own
+         // production round. In this case, we just use the hard deadline.
+         const auto hard_deadline = block_time.to_time_point() - fc::microseconds(chain::config::block_interval_us - cpu_effort_us);
+         return std::min(hard_deadline, now + fc::microseconds(cpu_effort_us));
+      }
+   }
+
+   inline uint32_t calculate_next_block_slot(const chain::account_name& producer_name, uint32_t current_block_slot, uint32_t block_num,
```

Comment:
I see why you made the return value `uint32_t`, but I think it would be cleaner if it returned a std::optional

Reply:
Refactored to not require an optional.